app/src/App.tsx

```diff
@@ -79,78 +57,156 @@ function App () {
     setToAmount(value)
   }, [])
 
+  const getTokenBalance = React.useCallback((token: BlockchainToken): string => {
+    return mockTokenBalances[token.contractAddress] ?? '0'
+  }, [mockTokenBalances])
+
+  // Needed to add this memo here to be declared before onClickPresetAmount
+  const fromTokenBalance = React.useMemo((): number => {
+    if (fromToken) {
+      return Number(getTokenBalance(fromToken))
+    }
+    return 0
+  }, [fromToken])
+
   const onClickPresetAmount = React.useCallback((key: 'half' | 'max') => {
-    const value = key === 'half' ? balance / 2 : balance
+    const value = key === 'half' ? fromTokenBalance / 2 : fromTokenBalance
     setFromAmount(value.toString())
-  }, [balance])
+  }, [fromTokenBalance])
 
   const onClickFlipSwapTokens = React.useCallback(() => {
     setFromToken(toToken)
     setToToken(fromToken)
   }, [toToken, fromToken])
 
   const onClickSettings = React.useCallback(() => {
-    // Logic here to open settings view
+    // Todo: Add logic here to open settings view
     return
   }, [])
 
+  const onClickReviewOrder = React.useCallback(() => {
+    // Todo: Add logic here to review order
+  }, [])
+
   const onOpenSelectToken = React.useCallback((id: 'to' | 'from') => () => {
-    if (id === 'from') {
-      setFromToken(mockEthereumToken)
+    setSelectingFromOrTo(id)
+    setShowSelectTokenModal(true)
+  }, [])
+
+  const onSelectToken = React.useCallback((token: BlockchainToken) => {
+    if (selectingFromOrTo === 'from') {
+      setFromToken(token)
+      setShowSelectTokenModal(false)
       return
     }
-    setToToken(mockWrappedSolanaToken)
+    setToToken(token)
+    setShowSelectTokenModal(false)
+  }, [selectingFromOrTo])
+
+  const onHideShowSelectTokenModal = React.useCallback(() => {
+    setShowSelectTokenModal(false)
   }, [])
 
+  // Effects
+  React.useEffect(() => {
+    if (toToken && fromToken && fromAmount !== '') {
+      setIsFetchingQuote(true)
+      setTimeout(() => {
+        const value = (Number(fromAmount) / rate).toFixed(6).toString()
+        setToAmount(value)
+        setIsFetchingQuote(false)
+      }, 1000)
+      return
+    }
+    setToAmount('')
+  }, [toToken, fromToken, fromAmount, rate])
+
+  // Memos
+  const fiatValue = React.useMemo((): string | undefined => {
+    if (fromAmount && price) {
+      return (price * Number(fromAmount)).toString()
+    }
+    return undefined
+  }, [price, fromAmount])
+
+  const insufficientBalance = React.useMemo((): boolean => {
+    if (fromAmount && fromTokenBalance !== undefined) {
+      return Number(fromAmount) > fromTokenBalance
+    }
+    return false
+  }, [fromTokenBalance, fromAmount])
+
+  const selectedToken = React.useMemo((): BlockchainToken | undefined => {
+    if (selectingFromOrTo === 'from') {
```

Comment:
Shouldn't this be reversed?

Reply:
No shouldn’t be reversed since we don’t want to allow selecting the same token for `from` if it’s already selected for `to`